### DISCUSSION TITLE
Cannot read properties of undefined (reading '_x_refs')

### DISCUSSION THREAD

**Discussion by dfallman** [alpinejs/alpine#3003] [1. Help] [answered]
All of a sudden I'm starting to get this every now and then... not sure where it comes from or what causes it (?)

```
Alpine Expression Error: Cannot read properties of undefined (reading '_x_refs')`
module.cjs.js:3391 Uncaught TypeError: Cannot read properties of undefined (reading '_x_refs')
    at handler2.inline (module.cjs.js:3391:1)
    at fullHandler (module.cjs.js:1797:1)
    at module.cjs.js:1983:1
    at Array.forEach (<anonymous>)
    at module.cjs.js:1983:1
    at walk (module.cjs.js:1912:1)
    at walk (module.cjs.js:1917:1)
    at walk (module.cjs.js:1917:1)
    at walk (module.cjs.js:1917:1)
    at walk (module.cjs.js:1917:1)
```

Any ideas? Understand I'm not giving a lot of info but everything seems to work as expected so wondering if it's a timing issue of some sort? Difficult/impossible to reproduce but happens (seemingly randomly) quite often.

**Comment by Triggs125** [comment, marked as answer]
This was happening to me as well. When I looked through the code in chrome dev tools I saw that a recursive function was looking through an element which had an `x-ref` property in it for any parent element which had an `x-data` property. Since there were no parent elements that matched this criteria, the recursive call eventually returned a null root element. The next piece of code tried to call `!root._x_refs` (seen in the error at module.cjs.js:3391) which would error because root is null. It seems like Alpine assumes that if `x-ref` is used then an `x-data` must be in a parent element.

The fix for me was to remove the `x-ref` property since it wasn't being used anyways (hence why this error didn't break anything). Another fix would probably be to add an `x-data` property in a parent element, but some more research into their docs might lead to another answer.

**Reply by dfallman** [reply]
Thanks -- will try this

**Reply by dfallman** [reply]
Our issue was subtly different but this led me onto finding the solution, thanks! 

In our case it seems we had timing issues with a fairly heavy page (including several embedded text editors, images, and bunch of other stuff on it) controlled by a `<template x-if="x == 1"...`. When `x` turns to 2 or something else a different but structurally identical page is inserted and the page corresponding to 1 gets removed from the DOM (or so we thought). 

I'm actually not sure why and/or if this is an Alpine issue, a browser issue, or just the way it is :) but it looks like template isn't able to remove the old page from the DOM quickly enough, so all of it (or at least parts of it) is still there, and when page 2 is injected into the DOM we ended up with several elements with the same x-ref. Not only did that result in the above console error, but also in a not insignificant memory leak situation.

Our solution was to first set `x = -1` (which in our system is equal to `no content page loaded`) and use `$nextTick()` to wait a click until x is set to the new value. This seems to give template enough time to fully remove the old page from the DOM and we no longer get this issue.

**Reply by josh-tt** [reply]
Thanks, I believe I'm having same issue here with swapping very large templates with x-if conditionals. Especially with fast clicking. Will try next tick solution, thanks.

**Reply by francescomugnai** [reply]
> Our issue was subtly different but this lead me onto finding the solution, thanks!
> 
> In our case it seems we had timing issues with a fairly heavy page (including several embedded text editors, images, and bunch of other stuff on it) controlled by a `<template x-if="x == 1"...`. When `x` turns to 2 or something else a different but structurally identical page is inserted and the page corresponding to 1 gets removed from the DOM (or so we thought).
> 
> I'm actually not sure why and/or if this is an Alpine issue, a browser issue, or just the way it is :) but it looks like template isn't able to remove the old page from the DOM quickly enough, so all of it (or at least parts of it) is still there, and when page 2 is injected into the DOM we ended up with several elements with the same x-ref. Not only did that result in the above console error, but also in a not insignificant memory leak situation.
> 
> Our solution was to first set `x = -1` (which in our system is equal to `no content page loaded`) and use `$nextTick()` to wait a click until x is set to the new value. This seems to give template enough time to fully remove the old page from the DOM and we no longer get this issue.

thank you! I had the same issue and I fixed it with $nextTick()

**Comment by christhekeele** [comment]
> This was happening to me as well. When I looked through the code in chrome dev tools I saw that a recursive function was looking through an element which had an x-ref property in it for any parent element which had an x-data property. Since there were no parent elements that matched this criteria, the recursive call eventually returned a null root element. 

This was happening to me too, in a new project where I really just wanted to use `$ref`, `$focus`, and `$store` I forgot to give *something* `x-data`. It feels like Alpine could provide a better error message here, is that worth raising a issue/PR?

**Reply by ekwoka** [reply]
Specifically that your had data wrapping something, but the ref was defined higher than that x-data?

That's a difficult situation to really detect with a useful message...

**Reply by christhekeele** [reply]
I was thinking more specifically that if we recurse to the null root, we could provide an error message saying "could not find enclosing x-data", rather than calling things like `root._x_refs` on it doomed to fail with an unhelpful message. That hint plush the stacktrace should give people what they need to solve the problem on their own.
